x/programs/cmd/simulator/src/lib.rs

```diff
@@ -95,17 +96,48 @@ pub enum Key {
 
 // TODO:
 // add `Cow` types for borrowing
-#[serde_as]
-#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
-#[serde(rename_all = "lowercase", tag = "type", content = "value")]
+#[derive(Clone, Debug, PartialEq)]
 pub enum Param {
-    U64(#[serde_as(as = "DisplayFromStr")] u64),
+    U64(u64),
     String(String),
     Id(Id),
-    #[serde(untagged)]
     Key(Key),
 }
 
+impl Serialize for Param {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut val = serializer.serialize_struct("param", 2)?;
+        match self {
+            Param::U64(num) => {
+                val.serialize_field("type", "u64")?;
+                val.serialize_field("value", &b64.encode(num.to_le_bytes()))?;
+            }
+            Param::String(text) => {
+                val.serialize_field("type", "string")?;
+                val.serialize_field("value", &b64.encode(text))?;
+            }
+            Param::Id(id) => {
+                val.serialize_field("type", "id")?;
+                let id = serde_json::to_vec(&id).map_err(serde::ser::Error::custom)?;
+                let id = &id[1..id.len() - 1]; // remove quotes
+                val.serialize_field("value", &b64.encode(id))?;
+            }
+            Param::Key(key) => {
+                let (key, algo) = match key {
+                    Key::Ed25519(key) => (key, "ed25519"),
+                    Key::Secp256r1(key) => (key, "secp256r1"),
+                };
+                val.serialize_field("type", algo)?;
+                val.serialize_field("value", &b64.encode(key))?;
+            }
+        }
+        val.end()
+    }
+}
```

Comment:
I don't quite understand why everything needs to be base64 encoded. If I had to guess, it's because it's easier to decode everything on the Go side, but you should just be able to *switch* on the `type` tag and only base64 decode where necessary. But IMO, it's an extra step that's unnecessary (and adds a little complexity). 

Also, I think it would just be beneficial to use an intermediate type here:
```rust
#[derive(Serialize)]
#[serde(rename_all = "lowercase", tag = "type", content = "value")]
enum StringParam {
    U64(String),
    String(String),
    Id(String),
    #[serde(untagged)]
    Key(Key),
}
```
If you want, you can define the type inside the `Serialize` implementation, do a `From` conversion, then just serialize the intermediate type. This should all get optimized away by the compiler, especially when the type is only used in one place. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=755c35cea6729ce0da44df255dceeaed

Reply:
It's Go making choices for us when dealing with the `[]byte` type, see https://pkg.go.dev/encoding/json#Marshal
> Array and slice values encode as JSON arrays, except that []byte encodes as a base64-encoded string, and a nil slice encodes as the null JSON value.
I found online that people were forking the package to strip out the base64 part but I don't think that we should do that.